test/new-e2e/pkg/environments/iis/component.go

```diff
@@ -0,0 +1,127 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package iis contains code for the IIS component
+package iis
+
+import (
+	"fmt"
+	"github.com/DataDog/datadog-agent/test/new-e2e/tests/windows/common/powershell"
+	"github.com/DataDog/test-infra-definitions/common"
+	"github.com/DataDog/test-infra-definitions/common/config"
+	"github.com/DataDog/test-infra-definitions/common/namer"
+	"github.com/DataDog/test-infra-definitions/common/utils"
+	infraComponents "github.com/DataDog/test-infra-definitions/components"
+	"github.com/DataDog/test-infra-definitions/components/command"
+	"github.com/DataDog/test-infra-definitions/components/remote"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+	"github.com/pulumiverse/pulumi-time/sdk/go/time"
+)
+
+// Output is an object that models the output of the resource creation
+// from the Component.
+// See https://www.pulumi.com/docs/concepts/resources/components/#registering-component-outputs
+type Output struct {
+	infraComponents.JSONImporter
+}
+
+// Component represent an IIServer and its sites.
+// See https://www.pulumi.com/docs/concepts/resources/components/
+type Component struct {
+	pulumi.ResourceState
+	infraComponents.Component
+	namer namer.Namer
+	host  *remote.Host
+}
+
+// Export registers a key and value pair with the current context's stack.
+func (c *Component) Export(ctx *pulumi.Context, out *Output) error {
+	return infraComponents.Export(ctx, c, out)
+}
+
+// NewServer creates a new IIS server component.
+func NewServer(ctx *pulumi.Context, e *config.CommonEnvironment, host *remote.Host, options ...Option) (*Component, error) {
+	params, err := common.ApplyOption(&Configuration{}, options)
+	if err != nil {
+		return nil, err
+	}
+
+	iisComponent, err := infraComponents.NewComponent(*e, host.Name(), func(comp *Component) error {
+		comp.namer = e.CommonNamer.WithPrefix(comp.Name())
+		comp.host = host
+
+		installIISCommand, err := host.OS.Runner().Command(comp.namer.ResourceName("install-iis"), &command.Args{
+			Create: pulumi.String(powershell.PsHost().
+				InstallIIS().
+				Compile()),
+		}, pulumi.Parent(comp))
+		if err != nil {
+			return err
+		}
+
+		timeProvider, err := time.NewProvider(ctx, comp.namer.ResourceName("time-provider"), &time.ProviderArgs{}, pulumi.DeletedWith(host))
+		if err != nil {
+			return err
+		}
+
+		waitForRebootCmd, err := time.NewSleep(ctx, comp.namer.ResourceName("wait-for-host-to-reboot"), &time.SleepArgs{
+			CreateDuration: pulumi.String("30s"),
+		}, pulumi.Provider(timeProvider), pulumi.DependsOn([]pulumi.Resource{
+			installIISCommand,
+		}))
+		if err != nil {
+			return err
+		}
+
+		ensureIISStarted, err := host.OS.Runner().Command(comp.namer.ResourceName("ensure-iis-started"), &command.Args{
+			Create: pulumi.String(powershell.PsHost().WaitForServiceStatus("W3SVC", "Running").Compile()),
+		}, pulumi.DependsOn([]pulumi.Resource{
+			waitForRebootCmd,
```

Comment:
Do I read this correctly that this will cause a reboot every time, even if a reboot is not required?
While it won't fail the test, it will make it take longer.

Also, it would be nice to be able to aggregate reboot requirements across environments.  I'm thinking of what we've discussed (in slack) separately, the need to be able to enable test-signed drivers.  These same tests can require both IIS and test-signed drivers.  Each may require a reboot; but you can reboot once at the end after doing both operations.

does this environment/provider interface allow for retaining state from one to the other, such that we can retain the reboot-required state and only reboot once at the end of the provisioning?

otherwise I can envision a more complex test setup where we're rebooting several times where only one is needed.